kubernetes/resource_kubernetes_certificate_signing_request.go

```diff
@@ -0,0 +1,204 @@
+	if err != nil {
+		return err
+	}
+
+	csr := v1beta1.CertificateSigningRequest{
+		ObjectMeta: metadata,
+		Spec:       *spec,
+	}
+	log.Printf("[INFO] Creating new certificate signing request: %#v", csr)
+	newCSR, createErr := conn.CertificatesV1beta1().CertificateSigningRequests().Create(&csr)
+	if createErr != nil {
+		return fmt.Errorf("Failed to create certificate signing request: %s", err)
+	}
+
+	// Get the name, since it might have been randomly generated during create.
+	csrName := newCSR.ObjectMeta.Name
+
+	if d.Get("auto_approve").(bool) {
+		retryErr := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+			pendingCSR, getErr := conn.CertificatesV1beta1().CertificateSigningRequests().Get(csrName, metav1.GetOptions{})
+			if getErr != nil {
+				return getErr
+			}
+			approval := v1beta1.CertificateSigningRequestCondition{
+				Type:    "Approved",
+				Reason:  "TerraformAutoApprove",
+				Message: "This CSR was approved by Terraform auto_approve.",
+			}
+			pendingCSR.Status.Certificate = []byte{}
+			pendingCSR.Status.Conditions = append(pendingCSR.Status.Conditions, approval)
+			_, updateErr := conn.CertificatesV1beta1().CertificateSigningRequests().UpdateApproval(pendingCSR)
+			return updateErr
+		})
+		if retryErr != nil {
+			panic(fmt.Errorf("CSR auto-approve update failed: %v", retryErr))
+		}
+		fmt.Println("CSR auto-approve update succeeded")
+		if err != nil {
+			return err
+		}
+	}
+
+	log.Printf("[DEBUG] Waiting for certificate to be issued")
+	stateConf := &resource.StateChangeConf{
+		Target:  []string{"Issued"},
+		Pending: []string{"", "Approved"},
+		Timeout: d.Timeout(schema.TimeoutCreate),
+		Refresh: func() (interface{}, string, error) {
+			out, refreshErr := conn.CertificatesV1beta1().CertificateSigningRequests().Get(csrName, metav1.GetOptions{})
+			if refreshErr != nil {
+				log.Printf("[ERROR] Received error: %v", refreshErr)
+				return out, "Error", refreshErr
+			}
+			var csrStatus string
+			emptyStatus := v1beta1.CertificateSigningRequestStatus{}
+			emptyCSR := v1beta1.CertificateSigningRequest{}
+
+			// Ensure the CSR fields have been populated before checking them.
+			if !reflect.DeepEqual(emptyCSR, out) {
+				if !reflect.DeepEqual(emptyStatus, out.Status) {
+					for _, condition := range out.Status.Conditions {
+						log.Printf("[DEBUG] Found Status.Condition.Type: %v", condition.Type)
+						if string(condition.Type) == "Approved" {
+							// Check to see if a certificate has been issued, and update status accordingly,
+							// since 'Issued' is not a state ever populated in the Status Conditions.
+							if string(out.Status.Certificate) != "" {
+								log.Printf("[DEBUG] Found non-empty Certificate field in Status")
+								csrStatus = "Issued"
+							}
+						}
+					}
+				}
+			}
+			log.Printf("[DEBUG] CertificateSigningRequest %s status received: %#v", csrName, csrStatus)
+			return out, csrStatus, nil
+		},
+	}
+	_, err = stateConf.WaitForState()
+	if err != nil {
```

Comment:
I hope I understand this workflow correctly. If I do, I think I see an issue here.

So if `auto_approve` is set to `false` then a human operator (or a manually triggered tool) will have to sign the request on the cluster cluster side. In that case, the timeout in the StateChangeConf is likely to expire and WaitForState will return an error. This will in turn fail the Create action and Terraform will fail the apply.

At this point, it seems to me there isn't any mechanism available to actually retrieve the (supposedly) eventually signed certificate. As I see it, the next possible action is only to run apply again, which will cause a new certificate request to be created (since the entire resource is marked ForceNew) and the entire timeout cycle to potentially repeat itself.

I wonder if actually implementing Read and Import can help here. Having a Read function would allow the Terraform operator to run a `refresh` operation and eventually retrieve the approved cert and reconcile the state to something stable.

Also, analyzing the chain of responsibility from the point of view of Terraform (and this provider), the only thing asked through the CSR resource is to issue a signing request (it's in the name of it too). Waiting for the actual signing to happen should be a best effort IMHO and not on the critical path with a failure mode attached.

I hope I didn't misunderstand how this works, but if I did please do set me on the straight path.

Reply:
Thanks for pointing this out! It actually showed me that I was making an assumption which turns out not to be true. So I'll fix this and test the timeout to make sure it works as expected. Here are some responses to specific parts of your comment to explain how it works:

> 
> So if `auto_approve` is set to `false` then a human operator (or a manually triggered tool) will have to sign the request on the cluster cluster side. In that case, the timeout in the StateChangeConf is likely to expire and WaitForState will return an error. This will in turn fail the Create action and Terraform will fail the apply.

The `auto_approve = false` setting is one that I tested by running `kubectl certificate approve` while the `terraform apply` was still running. This simulates the way a Kubernetes custom controller would quickly approve the request without manual intervention. But if needed, someone could also manually approve the CSR like I did. The most common use case though is a Kubernetes operator/controller.

> 
> At this point, it seems to me there isn't any mechanism available to actually retrieve the (supposedly) eventually signed certificate. 

This is why the `terraform apply` will wait until someone approves the request. If they don't approve it, then the apply will fail, and the remote resource will be deleted (or so I thought! but this is not true... yet).

I had never let it run for the full 5 minutes to see what would happen if it failed. I assumed if it failed then nothing would be saved in state, and so it would just try the create again... But when it does that, we get an error because the resource already exists in Kubernetes. This is the part I'll fix. What I need to do is defer the deletion of the API resource so that *even if it times out* the resource will be deleted, and the re-apply will be successful.

Reply:
It works now. Thanks!

```
kubernetes_certificate_signing_request.example: Still creating... [4m40s elapsed]
kubernetes_certificate_signing_request.example: Still creating... [4m50s elapsed]
kubernetes_certificate_signing_request.example: Still creating... [5m0s elapsed]

Error: timeout while waiting for state to become 'Issued' (timeout: 5m0s)

  on main.tf line 16, in resource "kubernetes_certificate_signing_request" "example":
  16: resource "kubernetes_certificate_signing_request" "example" {


[dakini@dax 20200806]$
[dakini@dax 20200806]$
[dakini@dax 20200806]$ tfapply
tls_private_key.example: Refreshing state... [id=413b4f1762fa9d7431cad39b15fccb7f131b4e72]
tls_cert_request.example: Refreshing state... [id=eeabc2dd57703ea3ba1002a511e60bf108f6f0ee]
kubernetes_certificate_signing_request.example: Creating...
kubernetes_certificate_signing_request.example: Still creating... [10s elapsed]
kubernetes_certificate_signing_request.example: Still creating... [20s elapsed]
kubernetes_certificate_signing_request.example: Creation complete after 23s [id=example]
kubernetes_secret.example: Creating...
kubernetes_secret.example: Creation complete after 0s [id=default/test-secret]
kubernetes_pod.main: Creating...
kubernetes_pod.main: Creation complete after 6s [id=default/test-pod]

Apply complete! Resources: 3 added, 0 changed, 0 destroyed.
```